[libc][tsearch] add tsearch functions

SchrodingerZhu opened the pull request:
closes #114695

This PR exposes the interal weak avl functionality to POSIX tsearch interfaces:

- tfind
- tsearch
- tdelete
- tdestroy
- twalk
- twalk_r

POSIX.1-2024 added  `typedef void posix_tnode`; this PR tries to align with latest POSIX specification.

SchrodingerZhu committed:
[libc][tsearch] add tsearch functions

SchrodingerZhu commented:
@michaelrj-google I have rebased the code onto its merged stack parent. PTAL.

One special place to pay additional attention to is the `__llvm_posix_tnode`. I am not very certain about the design. If you think this compatibility layer is not required, I can remove it.

SchrodingerZhu committed:
fix missed part from rebasing

michaelrj-google commented on libc/src/search/tdelete.cpp:
```diff
@@ -0,0 +1,36 @@
+    return nullptr;
+  void *result = const_cast<Node *>(node.value()->get_parent());
+  if (!result)
+    result = cpp::bit_cast<void *>(uintptr_t(-1));
```
The standard says the pointer is unspecified in this case. Since we're specifying it you should update teh list of unspecified behaviors in docs/dev/undefined_behavior

michaelrj-google reviewed (commented):
I got partway through reviewing but forgot to hit submit. Here's my comments:

SchrodingerZhu committed:
address CRs

michaelrj-google reviewed (commented):
The code looks fine, focusing on the tests for now

SchrodingerZhu commented on libc/test/src/search/tsearch_test.cpp:
```diff
@@ -0,0 +1,137 @@
+  int x = decode(a);
+  int y = decode(b);
+  return (x > y) - (x < y);
+}
```
I would think it is a common case but yeah, we don't handle allocation/deallocation case well. I will add an additional test.

SchrodingerZhu committed:
partially address the comments

SchrodingerZhu commented:
I only partially addressed the issues. I will take a look tomorrow when I got time.

SchrodingerZhu committed:
[libc] tsearch: rewrite tests

michaelrj-google reviewed (commented):
Some structural comments on the tests. They're mostly good but I think we can make them more useful.

SchrodingerZhu committed:
inline tests and add comments

the initial inlining is done by Claude Sonnet 4.6 based on existing code. Auditted by Schrodinger ZHU Yifan <i@zhuyi.fan>

SchrodingerZhu committed:
Adjust UB for nullfunction ptr to align with config

SchrodingerZhu commented:
PTAL. Sorry for the delay. Last few weeks were kind of crazy.

SchrodingerZhu committed:
Merge branch 'main' into libc/wavl-tsearch

michaelrj-google reviewed (approved):
LGTM, thanks for working on this! The tests look good now
